include/opendht/thread_pool.h

```diff
@@ -93,4 +93,78 @@ class OPENDHT_PUBLIC Executor : public std::enable_shared_from_this<Executor> {
     void schedule();
 };
 
+class OPENDHT_PUBLIC ExecutionContext {
+public:
+    ExecutionContext(ThreadPool& pool)
+     : threadPool_(pool), state_(std::make_shared<SharedState>())
+    {}
+
+    ~ExecutionContext() {
+        state_->destroy();
+    }
+
+    /** Wait for all scheduled tasks to be executed */
+    /*void await() {
+        state_->shutdown();
+    }*/
+
+    void run(std::function<void()>&& task) {
+        std::lock_guard<std::mutex> lock(state_->mtx);
+        if (state_->shutdown_) return;
+        state_->pendingTasks++;
+        threadPool_.get().run([task = std::move(task), state = state_] {
+            state->run(task);
+        });
+    }
+
+private:
+    struct SharedState {
+        std::mutex mtx {};
+        std::condition_variable cv {};
+        unsigned pendingTasks {0};
+        unsigned ongoingTasks {0};
+        /** When true, prevents new tasks to be scheduled */
+        bool shutdown_ {false};
+        /** When true, prevents scheduled tasks to be executed */
+        std::atomic_bool destroyed {false};
+
+        void shutdown() {
+            std::unique_lock<std::mutex> lock(mtx);
+            // wait for remaining tasks to complete
+            cv.wait(lock, [this] {
+                return pendingTasks == 0 && ongoingTasks == 0;
+            });
+            shutdown_ = true;
+        }
+
+        void destroy() {
+            std::unique_lock<std::mutex> lock(mtx);
+            shutdown_ = true;
+            cv.notify_all();
```

Comment:
Could you explain or leave a comment on why this notify_call is required?

Reply:
It was not necessary. I updated the PR.